libs/ml/include/ml/ops/ops.hpp

```diff
@@ -52,14 +54,34 @@ class Ops
    */
   virtual std::vector<SizeType> ComputeOutputShape(VecTensorType const &inputs) const = 0;
 
+  /**
+   * @brief ComputeBatchOutputShape is expensive function and should be used only for initialisation
+   * or in ASSERT. On Forward you can use output.shape() and on Backward there is
+   * error_signal.shape()
+   * @param input_shapes
+   * @return
+   */
+  virtual std::vector<SizeType> ComputeBatchOutputShape(ShapeVector const &input_shapes)
+  {
+    VecTensorType dummies;
```

Comment:
inefficient formulation. More efficient formulation is either to construct using input_shape.size() and then emplace back OR use vector.reserve() and then assign dummies[i] = ...
emplace back is probably neater

Reply:
done